ports/zephyr/machine_i2c.c

```diff
@@ -0,0 +1,145 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <zephyr.h>
+#include <i2c.h>
+
+#include "py/runtime.h"
+#include "py/gc.h"
+#include "py/mphal.h"
+#include "py/mperrno.h"
+#include "extmod/machine_i2c.h"
+
+STATIC const mp_obj_type_t machine_hard_i2c_type;
+
+typedef struct _machine_hard_i2c_obj_t {
+    mp_obj_base_t base;
+    struct device *dev;
+    bool restart;
+} machine_hard_i2c_obj_t;
+
+STATIC void machine_hard_i2c_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    machine_hard_i2c_obj_t *self = self_in;
+    mp_printf(print, "%s", self->dev->config->name);
+}
+
+mp_obj_t machine_hard_i2c_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    enum { ARG_id, ARG_scl, ARG_sda, ARG_freq, ARG_timeout };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_id, MP_ARG_REQUIRED | MP_ARG_OBJ },
+        { MP_QSTR_scl, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_sda, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_freq, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_timeout, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_obj = MP_OBJ_NULL} },
+    };
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all_kw_array(n_args, n_kw, all_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    const char *dev_name = mp_obj_str_get_str(args[ARG_id].u_obj);
+    struct device *dev = device_get_binding(dev_name);
+
+    if (!dev) {
+        mp_raise_ValueError("device not found");
+    }
+
+    if ((args[ARG_scl].u_obj != MP_OBJ_NULL) || (args[ARG_sda].u_obj != MP_OBJ_NULL)) {
+        mp_raise_NotImplementedError("explicit choice of scl/sda is not implemented");
+    }
+
+    if ((args[ARG_freq].u_obj != MP_OBJ_NULL)) {
+        mp_raise_NotImplementedError("explicit choice of freq is not implemented");
+    }
+
+    if ((args[ARG_timeout].u_obj != MP_OBJ_NULL)) {
+        mp_raise_NotImplementedError("explicit choice of timeout is not implemented");
+    }
+
+    machine_hard_i2c_obj_t *self = m_new_obj(machine_hard_i2c_obj_t);
+
+    self->base.type = &machine_hard_i2c_type;
+    self->dev = dev;
+    self->restart = false;
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+STATIC int machine_hard_i2c_transfer(mp_obj_base_t *self_in, uint16_t addr, uint8_t *buf, size_t len, bool stop, bool read) {
+    machine_hard_i2c_obj_t *self = (machine_hard_i2c_obj_t *)self_in;
+    struct i2c_msg msg;
+    int ret;
+
+    msg.buf = (u8_t *)buf;
```

Comment:
Unlike Zephyr, MicroPython codestyle avoids that strange-looking space in a cast, i.e. it should be `(u8_t*)`.